### PR TITLE
[Core] Adding hasher and comparator for IndexedObject (and pointer to IndexedObject)

### PR DESCRIPTION
**Pull request by loumalouomega**
Part of #5659

### CODE REVIEW

File: kratos/includes/key_hash.h [hunk 1]
```diff
@@ -219,6 +219,84 @@ namespace Kratos
...
+        {
+            return rIndexedObject.Id();
+        }
+    };
+
+    /**
+     * @brief This is a key comparer between two indexed objects
+     * @tparam TIndexedObject The type of indexed object to be compared
+     */
+    template<class TIndexedObject>
```
Comment (philbucher) [inline kratos/includes/key_hash.h]: why does it have to be a template if it is intended for the IndexedObject?

Comment (RiccardoRossi) [reply kratos/includes/key_hash.h]: i did not catch this...but indeed it should work without template for objects derived from IndexedObject

**Review by RiccardoRossi — COMMENTED** [review]
other than the template to me this is is ok

**Comment by loumalouomega** [comment]
I can remove the tamplate, but this works for classes not derived from IndexedObject but with an Id method (that probably should derive from IndexedObject instead)

**Comment by loumalouomega** [comment]
BTW, why indexed_object.h is in utilities, should be in includes

**Review by philbucher — APPROVED** [review]

**Pull request merged**